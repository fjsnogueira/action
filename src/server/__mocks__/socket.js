const socket = {
  emit: jest.fn(),
  setAuthToken: jest.fn()
};

export default socket;
