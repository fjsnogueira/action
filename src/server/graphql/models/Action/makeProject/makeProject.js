import getRethink from 'server/database/rethinkDriver';
import {
  GraphQLNonNull,
  GraphQLBoolean,
  GraphQLID
} from 'graphql';
import {requireSUOrTeamMember, requireWebsocket} from 'server/utils/authorization';

export default {
  type: GraphQLBoolean,
  description: 'Turn an action into a project',
  args: {
    actionId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The actionId (teamId::shortid) to delete'
    }
  },
  async resolve(source, {actionId}, {authToken, socket}) {
    const r = getRethink();

    // AUTH
    requireWebsocket(socket);
    const [teamId] = actionId.split('::');
    requireSUOrTeamMember(authToken, teamId);

    // RESOLUTION
    const action = await r.table('Action').get(actionId);
    const now = new Date();
    const newProject = {
      id: actionId,
      content: action.content,
      isArchived: false,
      teamId,
      teamMemberId: action.teamMemberId,
      createdAt: action.createdAt,
      updatedAt: now,
      status: 'active',
      sortOrder: 0,
      agendaId: action.agendaId
    };
    await r.table('Project').insert(newProject)
      .do(() => {
        return r.table('Action').get(actionId).delete();
      });
  }
};
