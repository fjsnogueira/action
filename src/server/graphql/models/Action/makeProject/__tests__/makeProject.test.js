import getRethink from 'server/database/rethinkDriver';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDate from 'mockdate';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import DynamicSerializer from 'dynamic-serializer';
import MockDB from 'server/__tests__/setup/MockDB';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import socket from 'server/__mocks__/socket';
import makeProject from 'server/graphql/models/Action/makeProject/makeProject';

MockDate.set(__now);
console.error = jest.fn();


describe('makeProject', () => {
  test('convert an action into a project', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();

    // TEST
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);
    await makeProject.resolve(undefined, {actionId}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      action: r.table('Action').get(actionId).default({}),
      project: r.table('Project').get(actionId)
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('throws when no websocket is present', async () => {
    const authToken = {};
    await expectAsyncToThrow(makeProject.resolve(undefined, {actionId: {}}, {authToken}));
  });

  test('throw when the caller is not a team member', async () => {
    // SETUP
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();

    // TEST
    const authToken = mockAuthToken(user[1], {tms: ['foo']});
    const actionId = action[0].id;

    // VERIFY
    await expectAsyncToThrow(makeProject.resolve(undefined, {actionId}, {authToken, socket}), [mockDB.context.team.id]);
  });
});
