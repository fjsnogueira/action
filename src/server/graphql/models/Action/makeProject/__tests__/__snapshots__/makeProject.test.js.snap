// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeProject convert an action into a project 1`] = `
Object {
  "action": Object {},
  "project": Object {
    "content": "Test Action[0]",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "id": 0,
    "isArchived": false,
    "sortOrder": 0,
    "status": "active",
    "teamId": 1,
    "teamMemberId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
  },
}
`;

exports[`makeProject throw when the caller is not a team member 1`] = `"{\\"_error\\":\\"You do not have access to team 0\\"}"`;

exports[`makeProject throws when no websocket is present 1`] = `"{\\"_error\\":\\"this must be called from a websocket\\"}"`;
